src/V3Udp.cpp

```diff
@@ -0,0 +1,235 @@
+// -*- mode: C++; c-file-style: "cc-mode" -*-
+//*************************************************************************
+// DESCRIPTION: Verilator: Implementation of User defined primitives
+//
+// Code available from: https://verilator.org
+//
+//*************************************************************************
+//
+// Copyright 2003-2025 by Wilson Snyder. This program is free software; you
+// can redistribute it and/or modify it under the terms of either the GNU
+// Lesser General Public License Version 3 or the Perl Artistic License
+// Version 2.0.
+// SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0
+//
+//*************************************************************************
+// V3Udp's Transformations:
+//
+// For every table line create an always block containing if statements
+// with condition depending on a combination of the input fields:
+//
+// 0 1 0 on a, b, c turns into !a&b&~c
+//
+//*************************************************************************
+
+#include "V3PchAstNoMT.h"  // VL_MT_DISABLED_CODE_UNIT
+
+#include "V3Udp.h"
+
+#include "V3Error.h"
+
+#include <vector>
+
+VL_DEFINE_DEBUG_FUNCTIONS;
+
+class UdpVisitor final : public VNVisitor {
+    bool m_inInitial = false;  // Is inside of an initial block
+    AstVar* m_oFieldVarp = nullptr;  // Output filed var of table line
+    std::vector<AstVar*> m_inputVars;  // All the input vars in the AstPrimitive
+    std::vector<AstVar*> m_outputVars;  // All the output vars in the AstPrimitive
+    AstPrimitive* m_primp = nullptr;  // The current primitive
+    bool m_isFirstOutput = false;  // Whether the first IO port is output
+    uint32_t m_inputNum = 0;  // The input var number
+    uint32_t m_outputNum = 0;  // The output var number
+    AstVarRef* m_outputInitVerfp = nullptr;
+    AstAlways* m_alwaysBlockp = nullptr;
+
+    void visit(AstInitial* nodep) override {
+        if (m_primp) m_inInitial = true;
+        iterateChildren(nodep);
+    }
+    void visit(AstAssign* nodep) override {
+        if (m_inInitial) { m_outputInitVerfp = VN_CAST(nodep->lhsp(), VarRef); }
+        iterateChildren(nodep);
+    }
+    void visit(AstPrimitive* nodep) override {
+        UASSERT_OBJ(!m_primp, nodep, "Primitives cannot nest");
+        VL_RESTORER(m_primp);
```

Comment:
Presumably every variable here should be VL_RESTORED.

While doesn't really matter given they can't nest I do occasionally look for missing VL_RESTORED's and makes it easier to understand intent.